File: objectivec-spritekit/Endless Runner/Endless Runner/view/Walls.m
//
//  Walls.m
//  Endless Runner
//
//  Created by Christopher Caleb on 17/04/2014.
//  Copyright (c) 2014 Christopher Caleb. All rights reserved.
//

#import "Walls.h"
#import "WallSlicePool.h"
#import "WallSlice.h"

@implementation Walls
{
    WallSlicePool *pool;
    NSMutableArray *slices;
    CGFloat viewportX;
    NSInteger viewportSliceX;
    SEL borrowWallSliceLookup[kSliceTypeGap];
    SEL returnWallSliceLookup[kSliceTypeGap];
    NSInteger VIEWPORT_NUM_SLICES;
}

- (id)initWithTextureAtlas:(SKTextureAtlas *)textureAtlas
{
    self = [super init];
    if (self)
    {
        VIEWPORT_NUM_SLICES = (NSInteger)ceil(VIEWPORT_WIDTH / WALL_SLICE_WIDTH) + 1;
        
        pool = [[WallSlicePool alloc] initWithTextureAtlas:textureAtlas];
        [self createLookupTables];
        
        slices = [NSMutableArray array];
        
        viewportX = 0;
        viewportSliceX = 0;
    }
    
    return self;
}

- (void)setViewportX:(CGFloat)newViewportX
{
    viewportX = [self checkViewportXBounds:newViewportX];
    NSInteger prevViewportSliceX = viewportSliceX;
    viewportSliceX = floor(viewportX / WALL_SLICE_WIDTH);
    [self removeOldSlices: prevViewportSliceX];
    [self addNewSlices];
}

- (void)addSlice:(SliceType)sliceType
{
    WallSlice *slice = [[WallSlice alloc] initWithType:sliceType andY:0.0f];
    [slices addObject:slice];
}

- (CGFloat)checkViewportXBounds:(CGFloat)newViewportX
{
    CGFloat maxViewportX = (slices.count - VIEWPORT_NUM_SLICES) * WALL_SLICE_WIDTH;
    if (newViewportX < 0)
    {
        newViewportX = 0;
    }
    else if (newViewportX > maxViewportX)
    {
        newViewportX = maxViewportX;
    }
    
    return newViewportX;
}

- (void)addNewSlices
{
    NSInteger firstX = -((NSInteger)viewportX % WALL_SLICE_WIDTH);
    for (NSInteger i = viewportSliceX, sliceIndex = 0;
                   i < viewportSliceX + VIEWPORT_NUM_SLICES;
                   i++, sliceIndex++)
    {
        WallSlice *slice = [slices objectAtIndex:i];
        if (slice.sprite == nil && slice.type != kSliceTypeGap)
        {
            slice.sprite = [self borrowWallSlice: slice.type];
            
            slice.sprite.position = CGPointMake(
                                                firstX + (sliceIndex * WALL_SLICE_WIDTH),
                                                slice.y);
            
            [self addChild:slice.sprite];
        }
        else if (slice.sprite != nil)
        {
            slice.sprite.position = CGPointMake(
                                                firstX + (sliceIndex * WALL_SLICE_WIDTH),
                                                slice.sprite.position.y);
        }
    }
}

- (void)removeOldSlices:(NSInteger)prevViewportSliceX
{
    NSInteger numOldSlices = viewportSliceX - prevViewportSliceX;
    if (numOldSlices > VIEWPORT_NUM_SLICES)
    {
        numOldSlices = VIEWPORT_NUM_SLICES;
    }
    
    for (NSInteger i = prevViewportSliceX; i < prevViewportSliceX + numOldSlices; i++)
    {
        WallSlice *slice = slices[i];
        if (slice.sprite != nil)
        {
            [self returnWallSlice:slice.type usingSprite:slice.sprite];
            [slice.sprite removeFromParent];
            slice.sprite = nil;
        }
    }
}

- (void)addSlice:(SliceType)sliceType atYPosition:(CGFloat)y
{
    WallSlice *slice = [[WallSlice alloc] initWithType:sliceType andY:y];
    [slices addObject:slice];
}

- (void)createLookupTables
{
    borrowWallSliceLookup[kSliceTypeFront] = @selector(borrowFrontEdge);
    borrowWallSliceLookup[kSliceTypeBack] = @selector(borrowBackEdge);
    borrowWallSliceLookup[kSliceTypeStep] = @selector(borrowStep);
    borrowWallSliceLookup[kSliceTypeDecoration] = @selector(borrowDecoration);
    borrowWallSliceLookup[kSliceTypeWindow] = @selector(borrowWindow);
    
    returnWallSliceLookup[kSliceTypeFront] = @selector(returnFrontEdge:);
    returnWallSliceLookup[kSliceTypeBack] = @selector(returnBackEdge:);
    returnWallSliceLookup[kSliceTypeStep] = @selector(returnStep:);
    returnWallSliceLookup[kSliceTypeDecoration] = @selector(returnDecoration:);
    returnWallSliceLookup[kSliceTypeWindow] = @selector(returnWindow:);
}

- (SKSpriteNode *)borrowWallSlice:(SliceType)sliceType
{
    return [pool performSelector:borrowWallSliceLookup[sliceType]];
}

- (void)returnWallSlice:(SliceType)sliceType usingSprite:(SKSpriteNode *)sprite
{
    [pool performSelector:returnWallSliceLookup[sliceType] withObject:sprite];
}
@end

File: swift-spritekit/Endless-Runner/Walls.swift
//
//  Walls.swift
//  Endless-Runner
//
//  Created by Christopher Caleb on 13/06/2014.
//  Copyright (c) 2014 Christopher Caleb. All rights reserved.
//

import Foundation
import SpriteKit

class Walls: SKNode {
    let VIEWPORT_WIDTH: CGFloat = 1024.0
    let VIEWPORT_NUM_SLICES: NSInteger
    
    let _pool: WallSlicePool
    var _slices: WallSlice[]
    var _viewportX: CGFloat
    var _viewportSliceX: Int
    var _borrowWallSliceLookup: (() -> SKSpriteNode)[] = []
    var _returnWallSliceLookup: ((SKSpriteNode) -> ())[] = []
    
    init(textureAtlas: SKTextureAtlas) {
        VIEWPORT_NUM_SLICES = NSInteger(ceil(CDouble(VIEWPORT_WIDTH / WALL_SLICE_WIDTH))) + 1
        
        _pool = WallSlicePool(textureAtlas: textureAtlas)
        
        _slices = []
        
        _viewportX = 0
        _viewportSliceX = 0
        
        super.init()
        
        createLookupTables()
    }
    
    func setViewportX(newViewportX: CGFloat) {
        _viewportX = checkViewportXBounds(newViewportX)
        let prevViewportSliceX = _viewportSliceX
        _viewportSliceX = NSInteger(floor(CDouble(_viewportX) / CDouble(WALL_SLICE_WIDTH)))
        removeOldSlices(prevViewportSliceX)
        addNewSlices()
    }
    
    func addSlice(sliceType: SliceType, y: CGFloat = 0) {
        let slice = WallSlice(type: sliceType, y: y)
        _slices.append(slice)
    }
    
    func checkViewportXBounds(var newViewportX: CGFloat) -> CGFloat {
        var maxViewportX = CGFloat(_slices.count - VIEWPORT_NUM_SLICES) * WALL_SLICE_WIDTH
        if newViewportX < 0 {
            newViewportX = 0
        } else if newViewportX > maxViewportX {
            newViewportX = maxViewportX
        }
        
        return newViewportX
    }
    
    func addNewSlices() {
        let firstX :NSInteger = -NSInteger(_viewportX % WALL_SLICE_WIDTH)
        for var i: NSInteger = _viewportSliceX, sliceIndex = 0;
            i < _viewportSliceX + VIEWPORT_NUM_SLICES;
            i++, sliceIndex++ {
                let slice = _slices[i]
                if slice.sprite == nil && slice.type != .Gap {
                    slice.sprite = borrowWallSlice(slice.type)
                    if let sprite = slice.sprite {
                        let x = CGFloat(firstX) + (CGFloat(sliceIndex) * WALL_SLICE_WIDTH)
                        sprite.position = CGPoint(x: x, y: slice.y)
                    }
                    addChild(slice.sprite)
                } else if slice.sprite != nil {
                    if let sprite = slice.sprite {
                        let x = CGFloat(firstX) + (CGFloat(sliceIndex) * WALL_SLICE_WIDTH)
                        let y = sprite.position.y
                        sprite.position = CGPoint(x: x, y: y)
                }
            }
        }
    }
    
    func removeOldSlices(prevViewportSliceX: NSInteger) {
        var numOldSlices = _viewportSliceX - prevViewportSliceX
        if numOldSlices > VIEWPORT_NUM_SLICES {
            numOldSlices = VIEWPORT_NUM_SLICES
        }
        
        for i in prevViewportSliceX..(prevViewportSliceX + numOldSlices) {
            let slice = _slices[i]
            if let sprite = slice.sprite {
                returnWallSlice(slice.type, sprite: sprite)
                sprite.removeFromParent()
                slice.sprite = nil
            }
        }
    }
    
    func createLookupTables() {
        _borrowWallSliceLookup = Array(count:5, repeatedValue:_pool.borrowWindow)
        _borrowWallSliceLookup[SliceType.Front.toRaw()] = _pool.borrowFrontEdge
        _borrowWallSliceLookup[SliceType.Back.toRaw()] = _pool.borrowBackEdge
        _borrowWallSliceLookup[SliceType.Step.toRaw()] = _pool.borrowStep
        _borrowWallSliceLookup[SliceType.Decoration.toRaw()] = _pool.borrowDecoration
        _borrowWallSliceLookup[SliceType.Window.toRaw()] = _pool.borrowWindow
        
        _returnWallSliceLookup = Array(count:5, repeatedValue:_pool.returnWindow)
        _returnWallSliceLookup[SliceType.Front.toRaw()] = _pool.returnFrontEdge
        _returnWallSliceLookup[SliceType.Back.toRaw()] = _pool.returnBackEdge
        _returnWallSliceLookup[SliceType.Step.toRaw()] = _pool.returnStep
        _returnWallSliceLookup[SliceType.Decoration.toRaw()] = _pool.returnDecoration
        _returnWallSliceLookup[SliceType.Window.toRaw()] = _pool.returnWindow
    }
    
    func borrowWallSlice(sliceType: SliceType) -> SKSpriteNode {
        return _borrowWallSliceLookup[sliceType.toRaw()]()
    }
    
    func returnWallSlice(sliceType: SliceType, sprite: SKSpriteNode) {
        _returnWallSliceLookup[sliceType.toRaw()](sprite)
    }
}